lib/Extension/LanguageServerCodeTransform/CodeAction/GenerateMutatorsProvider.php

```diff
@@ -0,0 +1,90 @@
+<?php
+
+namespace Phpactor\Extension\LanguageServerCodeTransform\CodeAction;
+
+use Amp\CancellationToken;
+use Amp\Promise;
+use Phpactor\Extension\LanguageServerBridge\Converter\PositionConverter;
+use Phpactor\Extension\LanguageServerCodeTransform\LspCommand\GenerateMutatorsCommand;
+use Phpactor\LanguageServerProtocol\CodeAction;
+use Phpactor\LanguageServerProtocol\Command;
+use Phpactor\LanguageServerProtocol\Range;
+use Phpactor\LanguageServerProtocol\TextDocumentItem;
+use Phpactor\LanguageServer\Core\CodeAction\CodeActionProvider;
+use Phpactor\WorseReflection\Core\Reflection\ReflectionClass;
+use Phpactor\WorseReflection\Core\Reflection\ReflectionProperty;
+use Phpactor\WorseReflection\Reflector;
+use function Amp\call;
+
+class GenerateMutatorsProvider implements CodeActionProvider
+{
+    public const KIND = 'quickfix.generate_mutators';
+
+    private Reflector $reflector;
+
+    public function __construct(Reflector $reflector)
+    {
+        $this->reflector = $reflector;
+    }
+
+    public function kinds(): array
+    {
+        return [
+             self::KIND
+         ];
+    }
+
+    public function provideActionsFor(TextDocumentItem $textDocument, Range $range, CancellationToken $cancel): Promise
+    {
+        return call(function () use ($range, $textDocument) {
+            // CoC will select the entire docyment if no range selected
+            if ($range->start->line === 0 && $range->start->character === 0) {
+                return [];
+            }
+            $startOffset = PositionConverter::positionToByteOffset($range->start, $textDocument->text)->toInt();
+            $endOffset = PositionConverter::positionToByteOffset($range->end, $textDocument->text)->toInt();
+
+            $classes = $this->reflector->reflectClassesIn($textDocument->text);
+
+            if ($classes->count() === 0) {
+                return [];
+            }
+
+            // TODO: Class at offset
+            $reflectionClass = $classes->first();
+
+            if (!$reflectionClass instanceof ReflectionClass) {
+                return [];
+            }
+
+            $propertyNames = [];
+            foreach ($reflectionClass->properties() as $property) {
+                assert($property instanceof ReflectionProperty);
+                if ($property->position()->start() < $startOffset || $property->position()->end() > $endOffset) {
+                    continue;
+                }
+                $propertyNames[] = $property->name();
+            }
+
+            if (empty($propertyNames)) {
+                return [];
+            }
+
+            return [
+                CodeAction::fromArray([
+                    'title' => sprintf('Generate %s mutator(s)', count($propertyNames)),
+                    'kind' => self::KIND,
+                    'command' => new Command(
+                        sprintf('Generate %s mutator(s)', count($propertyNames)),
+                        GenerateMutatorsCommand::NAME,
```

Comment:
is there anything different about this command other than the command name? can we parameterise the Accessor class rather than copy / paste this one?

Reply:
They're pretty much the same. I refactored so we only have one provider now. As you can see in the class's constructor, we have 3 new arguments (`kind`, `command`, and `type`). However, I'm not satisfied with the name of the `type` property, maybe you have a better suggestion?

See https://github.com/phpactor/phpactor/pull/1946/files#diff-86a88fb7dd73a00bf12d3e88a9c9bf24a70729458439176bd56afef0775dce7cR31